Google Pay Support

deivamagalhaes opened the pull request:
# Summary

This PR contains the FW changes to add support to Google Pay.

### Story: [CH 65740](https://app.clubhouse.io/skyverge/story/65740/add-framework-support-to-google-pay)
### Release: #510

## Details

The JS is based on [this tutorial](https://developers.google.com/pay/api/web/guides/tutorial#full-example).

The reason why we had to block and unblock the UI on every step of the process is that Google Pay JS does not provide a cancellation callback, so if we don't unblock the UI and the user cancels the payment/close the pop-up, the UI remains blocked forever. 😞 

The backend is roughly based on Apple Pay implementation. The shared code was moved to new parent abstract classes.

## UI Changes

- New Google Pay settings page (similar to Apple Pay): https://cloud.skyver.ge/eDuPBzOn
- New Google Pay button: https://cloud.skyver.ge/nOunPLmy

## QA

### Setup

1. Enable Subscriptions
1. Configure a shipping zone with at least one paid shipping method
1. Navigate to your `wc-plugins` repository folder
1. Checkout branch `ch61993/cybersource-google-pay` 
1. Navigate to your CyberSource folder
1. Run `composer install`
1. Activate and configure CyberSource
1. Navigate to WooCommerce > Settings > Payments > Google Pay
1. Enable Google Pay
1. Enable to show on all pages (Product, Cart and Checkout)
1. Enable Test Mode (so that we don't test real CyberSource stuff, just the FW stuff)

### Steps

#### Buy from Product Page

1. Navigate to a simple product page
    - [x] The Google Pay button is displayed
1. Click the Google Pay button
    - [x] The Google Pay popup is displayed and allows you to provide a shipping address
1. Sign in to your Google account (if needed) and provide a valid credit card (it won't be charged) and a shipping address
1. Click "Pay"
    - [x] The order is placed successfully with the correct shipping and billing info and shipping costs

#### Buy from Cart Page

1. Add a simple product to your cart
1. Navigate to the cart
    - [x] The Google Pay button is displayed
1. Click the Google Pay button
    - [x] The Google Pay popup is displayed and allows you to provide a shipping address
1. Sign in to your Google account (if needed) and provide a valid credit card (it won't be charged) and a shipping address
1. Click "Pay"
    - [x] The order is placed successfully with the correct shipping and billing info and shipping costs

#### Buy from Checkout Page

1. Add a simple product to your cart
1. Navigate to the Checkout page
    - [x] The Google Pay button is displayed
1. Click the Google Pay button
    - [x] The Google Pay popup is displayed and allows you to provide a shipping address
1. Sign in to your Google account (if needed) and provide a valid credit card (it won't be charged) and a shipping address
1. Click "Pay"
    - [x] The order is placed successfully with the correct shipping and billing info and shipping costs

#### Virtual product from the Product page

1. Navigate to a virtual product page
    - [x] The Google Pay button is displayed
1. Click the Google Pay button
    - [x] The Google Pay popup is displayed and does not allow you to provide a shipping address
1. Sign in to your Google account (if needed) and provide a valid credit card (it won't be charged)
1. Click "Pay"
    - [ ] The order is placed successfully with the correct billing info

#### Virtual product from the Cart page

1. Add a virtual product to your cart
1. Navigate to the cart
    - [x] The Google Pay button is displayed
1. Click the Google Pay button
    - [x] The Google Pay popup is displayed and does not allow you to provide a shipping address
1. Sign in to your Google account (if needed) and provide a valid credit card (it won't be charged)
1. Click "Pay"
    - [x] The order is placed successfully with the correct billing info

#### Virtual product from the Checkout page

1. Add a virtual product to your cart
1. Navigate to the Checkout page
    - [x] The Google Pay button is displayed
1. Click the Google Pay button
    - [x] The Google Pay popup is displayed and does not allow you to provide a shipping address
1. Sign in to your Google account (if needed) and provide a valid credit card (it won't be charged)
1. Click "Pay"
    - [x] The order is placed successfully with the correct billing info

#### Cancel payment

1. Navigate to a simple product page
    - [x] The Google Pay button is displayed
1. Click the Google Pay button
    - [x] The Google Pay popup is displayed and allows you to provide a shipping address
1. Close the Google Pay popup
    - [x] The UI is unblocked and you can try another payment method

#### Unsupported shipping address

1. Navigate to a simple product page
    - [x] The Google Pay button is displayed
1. Click the Google Pay button
    - [x] The Google Pay popup is displayed and allows you to provide a shipping address
1. Sign in to your Google account (if needed) and provide a valid credit card (it won't be charged) and a shipping address outside of your shipping zone
    - [x] An error message saying "Cannot ship to the selected address" is displayed and you cannot move forward with the payment

#### Subscription product

1. Navigate to a subscription product page
    - [x] The Google Pay button is not displayed
1. Add the subscription to the cart
1. Navigate to the cart
    - [x] The Google Pay button is not displayed
1. Navigate to the Checkout page
    - [x] The Google Pay button is not displayed
1. Navigate to a simple product page
    - [x] The Google Pay button is displayed
1. Click the Google Pay button
    - [ ] An error message is displayed informing there is a subscription on the cart

#### Button style

1. Navigate to WooCommerce > Settings > Payments > Google Pay
1. Change the button style to white
1. Navigate to a simple product page
    - [x] The white Google Pay button is displayed

#### Terms

1. Create a terms page and configure it in WooCommerce > Settings > Advanced
1. Navigate to a simple product page
    - [x] A link to the terms page is displayed below the Google Pay button

#### Apple Pay (backwards compatibility)

1. Make a purchase using Apple Pay from a product page
    - [x] The order is placed successfully 
1. Make a purchase using Apple Pay from the cart page
    - [x] The order is placed successfully 
1. Make a purchase using Apple Pay from the checkout page
    - [x] The order is placed successfully 

## Before merge

- [ ] I have confirmed these changes in each supported minor WooCommerce version

N-Molham reviewed (approved):
Looking good

deivamagalhaes commented:
@skyverge/engineering-wordpress, I'd appreciate one more set of eyes on this one, since it is such a huge PR.

ChaseWiseman reviewed (changes requested):
@deivamagalhaes QA steps pass! This is coming along nicely. Aside from code review, some initial comments:

## General
- I see this is enabled only via a filter - is that a requirement of the CyberSource part of this? Apple Pay requires a lengthy dev-focused setup process, so it's a dev-only feature behind its filter, but in this case I don't see any additional configuration steps. Is there a reason to enforce that for Google Pay? If not, I think we could just rely on the plugin/gateway having the feature flag constant.
- Is their flow limited to shippable physical products? Currently if I try it with a virtual product it still attempts to use a shipping address and the payment can't be completed.
- It looks like we could use some styling for the new button, as right now it's scrunched in with other UI:
    - On the product page: https://cloud.skyver.ge/12urvWR7
    - On the cart page (we could also probably make that full container width if they allow it, to match Apple Pay and the Checkout button): https://cloud.skyver.ge/9ZuBRl9d
    - On the checkout page there's no styling around the button or the "or": https://cloud.skyver.ge/2NujPYp0 

## Apple Pay
Apart from the issues above, in isolation this feature works well. However when Apple Pay is also enabled on a site, which I think we can expect will be the case for a number of merchants, there are some notable UI clashes.

- The "terms" UI below the buttons is duplicated (product, cart, or checkout): https://cloud.skyver.ge/Wnurk2vX
- Both features try and add their own "or" UI to Checkout: https://cloud.skyver.ge/Kouw70bv

It seems to me like we'll need to find a common abstraction of some of these things, rather than duplicating the UI handling code for each. Especially for something like the `Orders` class, which doesn't seem to contain any code specific to Google Pay, all of these new features can probably benefit from shared code now that we're starting to add more than just Apple Pay.

cc @wvega here too, as I'm assuming the Visa Checkout feature will run into the same situation if it's also a "buy now" button type of feature.

What do y'all think?

deivamagalhaes commented:
> It seems to me like we'll need to find a common abstraction of some of these things, rather than duplicating the UI handling code for each. Especially for something like the Orders class, which doesn't seem to contain any code specific to Google Pay, all of these new features can probably benefit from shared code now that we're starting to add more than just Apple Pay.

@ChaseWiseman any suggestion on a name for this general handler that would contain both Google Pay and Apple Pay stuff? `External_Checkout` maybe? 

And how do we want the UI to look when we have both enabled?

[Google Pay brand guidelines](https://developers.google.com/pay/api/web/guides/brand-guidelines) allow to place the button next to another button, as long as they are the same size:
> If you place a Google Pay button next to another button, make sure the Google Pay button is of equal size.

[Apple Pay Marketing Guidelines](https://developer.apple.com/apple-pay/marketing/) mention the same thing:
> Do not display a mark that is smaller than other payment identities in a similar format.

So I think it should be okay to display both buttons next to each other and the terms below the buttons (when applicable). Thoughts?

wvega commented:
@deivamagalhaes @ChaseWiseman for the Orders handler I was thinking something like `Framework\Payment_Gateway\Checkout\Orders.php`.

I'm using this implementation https://gist.github.com/wvega/0a59dac11d166973d7dc5875b037547d for the Visa Checkout changes. The only difference with the Apple Pay Orders handler is that `create_order()` takes an array of `$order_data` so that users can set the `created_via` property. I haven't updated Apple Pay to use that one though.

I was also considering `Framework\Payment_Gateway\Settings_Screen` as a base handler for the admin UI. A version of those changes is here: https://github.com/skyverge/wc-plugin-framework/pull/508

The names are suggestions of course. I'm happy to change them and update the code I'm working on to accommodate what we decide here.

wvega commented:
As for other ideas, while working on Visa Checkout I started thinking of Apple Pay, Google Pay, and others like it as "Payment Gateway extensions", so maybe some of those common helpers can live under the `Framework\Payment_Gateway\Extensions` namespace.

deivamagalhaes commented:
@wvega, from your Gist:

> * A checkout helper that handles order creation outside the regular checkout process. 

I think if we can find a name that has the same meaning as that sentence, it would be perfect. `Extensions` feels a bit too vague to me. 

That is why I thought of `External_Checkout`. But maybe someone that has English as a first language has a better idea. 😬

deivamagalhaes commented:
@ChaseWiseman , I've moved all Apple Pay and Google Pay files into a new `External_Checkout` namespace and extracted the common Orders handler to it as well. But I am not sure it's a good idea to do this refactor while Will is also working on Visa Checkout. Maybe it would be better to merge as is and do the refactor on the release branch instead? Just thinking about conflicts...

deivamagalhaes commented:
> Is their flow limited to shippable physical products? Currently, if I try it with a virtual product it still attempts to use a shipping address and the payment can't be completed.

@ChaseWiseman, I missed that. I guess I will have to update the JS to use the flow that does not require a shipping address for virtual products. 🤔

ChaseWiseman commented:
>So I think it should be okay to display both buttons next to each other and the terms below the buttons (when applicable). 

@deivamagalhaes @wvega sounds fine to me, just as long as whatever styling you land on considers that there my be more than 2 buttons, e.g. once Visa Checkout is also added.

>Maybe it would be better to merge as is and do the refactor on the release branch instead?

I think ideally there would be an agreed upon implementation of the abstracted component, then this branch and the Visa Checkout branch would work off of that to extend those components before we merge them.

It sounds like some additional coordinated discovery is needed to finalize such an approach, while also considering the scope of both of the epics and their timelines and triggering the types of conversations that need to happen if scope gets too large. For instance, do the drawbacks mentioned above (duplicate code & UI elements for each button) outweigh the extra discovery and development that'll be needed?

deivamagalhaes commented:
@ChaseWiseman, I've:
- fixed the behaviour for virtual products 
- tweaked the button style
- with @wvega's help (thanks 💜), refactored the code so that Google Pay and Apple Pay share a parent main handler, Admin handler, and Frontend handler with their common code 😅 

Can you please take another look? It's probably a good idea to redo the QA as well... 😬 

I am also trying to figure out why I'm getting 404 for the JS map files. Shouldn't it be trying to get them from the same path as the JS files? @alexanderchan or @wesleycole, ideas?

deivamagalhaes commented:
I just noticed I also need to tweak the Apple Pay style, because it is now showing an invisible button on Chrome:
https://cloud.skyver.ge/DOu9dAYv

🤔

deivamagalhaes commented:
Fixed the Apple Pay style issue. It now hides and shows the button (and terms, when appropriate) as before.

deivamagalhaes commented:
I also noticed an issue with Checkout Add-ons. If you add an add-on with a cost, this cost is not added to the Apple Pay or Google Pay total. Since this is an existing and localized issue, I believe we should open a separate story for it.

deivamagalhaes commented:
404 on JS map files also fixed. ✅

deivamagalhaes commented:
Stories created for the [Checkout Add-on issue](https://github.com/skyverge/wc-plugin-framework/pull/507#issuecomment-715503988): 
[CH 66364](https://app.clubhouse.io/skyverge/story/66364/checkout-add-on-cost-is-not-added-to-apple-pay-total)
[CH 66368](https://app.clubhouse.io/skyverge/story/66368/checkout-add-on-cost-is-not-added-to-google-pay-total)

ChaseWiseman reviewed (changes requested):
@deivamagalhaes looking good! Nice work on all of that abstraction 🍻 

A few QA issues to address then this should be mergeable:

These all seem to be present regardless of whether Apple Pay is available or enabled. I replicated these after running `composer install` on the latest in the Cybs branch.

### Buy from Product Page
- [x] The shipping methods load as options, but the selected method doesn’t seem to come through as a line item & isn’t added to the total. The order ends up being the correct total w/ shipping
- [ ] Taxes are also not applied in the Google UI

#### Example log
```
10-23-2020 @ 21:56:11 - [Google Pay] Recalculating totals
10-23-2020 @ 21:56:11 - [Google Pay] New totals:
Array
(
    [newTransactionInfo] => Array
        (
            [displayItems] => Array
                (
                    [0] => Array
                        (
                            [label] => Subtotal
                            [type] => SUBTOTAL
                            [price] => 1.00
                        )

                )

            [countryCode] => US
            [currencyCode] => USD
            [totalPriceStatus] => FINAL
            [totalPrice] => 1.00
            [totalPriceLabel] => Total
        )

    [newShippingOptionParameters] => Array
        (
            [shippingOptions] => Array
                (
                    [0] => Array
                        (
                            [id] => flat_rate:1
                            [label] => Super Fast
                            [description] => 
                        )

                    [1] => Array
                        (
                            [id] => flat_rate:2
                            [label] => Super Slow
                            [description] => 
                        )

                    [2] => Array
                        (
                            [id] => free_shipping:3
                            [label] => Free shipping
                            [description] => 
                        )

                    [3] => Array
                        (
                            [id] => local_pickup:4
                            [label] => Local pickup
                            [description] => 
                        )

                )

            [defaultSelectedOptionId] => flat_rate:2
        )

)
```

### Virtual product from the Product page
- [x] The order is created with no line items and a $0 total
- [x] A PHP notice is triggered: `- There’s a notice: PHP Notice:  Undefined property: stdClass::$shippingAddress in vendor/skyverge/wc-plugin-framework/woocommerce/payment-gateway/External_Checkout/Google_Pay/Google_Pay.php on line 438`

### Buy from Cart Page
- [ ] If I change my shipping method in the Woo UI, the buttons disappear and don’t come back
- [ ] There appears to be an Apple Pay request logged (if enabled), but weirdly it logs as a Google request

```
10-23-2020 @ 21:59:16 - [Google Pay] Getting payment request
10-23-2020 @ 21:59:16 - [Google Pay] Payment Request:
Array
(
    [currencyCode] => USD
    [countryCode] => US
    [merchantCapabilities] => Array
        (
            [0] => supports3DS
            [1] => supportsCredit
            [2] => supportsDebit
        )

    [supportedNetworks] => Array
        (
            [0] => amex
            [1] => discover
            [2] => masterCard
            [3] => visa
        )
```

deivamagalhaes commented on woocommerce/payment-gateway/class-sv-wc-payment-gateway.php:
```diff
@@ -509,13 +512,13 @@ protected function get_payment_form_js_localized_script_params() {
 
 		$handle = $this->get_gateway_js_handle();
-		$js_path   = $this->get_plugin()->get_plugin_path() . '/assets/js/frontend/' . $handle . '.min.js';
+		$js_path   = $this->get_plugin()->get_plugin_path() . '/assets/dist/frontend/' . $handle . '.js';
```
Not sure I understood this one. We are not changing the path for the plugin JS files, just the gateway. So I don't think we will have a `$this->get_plugin()->get_plugin_path() . '/assets/dist/frontend/'` folder.

ChaseWiseman replied:
>So I don't think we will have a $this->get_plugin()->get_plugin_path() . '/assets/dist/frontend/' folder.

@deivamagalhaes yep that's right. that dir doesn't exist in the plugins, so their JS won't be enqueued properly. While this method is in the gateway class, it applies somewhat at a plugin level unless overridden by individual gateways. Most of them have a single plugin-level JS file that is used by all of the gateways that it has.

Using Cybersource as an example, its main JS file is `woocommerce-gateway-cybersource/assets/js/frontend/wc-cybersource.min.js`. The plugin itself doesn't enqueue this file - the framework does that for it in this method.

With this change, the framework will now look for that JS file at `woocommerce-gateway-cybersource/assets/dist/frontend/wc-cybersource.js` and won't find it because plugins haven't adopted the framework's new compiling structure. On the accompanying Cybersource branch, I can see that the JS for the flex form is no longer enqueued.

So the `$js_path` and `$js_url` values need to remain unchanged to maintain backwards compatibility with how plugins have their JS structured currently. Does that make sense?

deivamagalhaes replied:
I thought you were asking me to change that line, instead of telling me not to change it. 🤦‍♀️  

We should also keep the `.min` suffix, right?

deivamagalhaes replied:
Fixed in https://github.com/skyverge/wc-plugin-framework/pull/507/commits/3cba64529d90c7eccd98c495c7546dba563524c3.

deivamagalhaes commented on woocommerce/payment-gateway/External_Checkout/Frontend.php:
```diff
@@ -0,0 +1,428 @@
+		// add the action and filter hooks
+		$this->add_hooks();
+
+		parent::__construct();
```
Fixed in https://github.com/skyverge/wc-plugin-framework/pull/507/commits/bb3c644acd6244db323b6293a4bc218120600ade

deivamagalhaes commented on woocommerce/payment-gateway/External_Checkout/Frontend.php:
```diff
@@ -0,0 +1,428 @@
+	 *
+	 * @since 5.10.0
+	 *
+	 * @return Google_Pay|\SkyVerge\WooCommerce\PluginFramework\v5_10_0\SV_WC_Payment_Gateway_Apple_Pay
```
Fixed in https://github.com/skyverge/wc-plugin-framework/pull/507/commits/8a5fca6ca2fb979e460157c3ed311d465720407c

deivamagalhaes commented on woocommerce/payment-gateway/External_Checkout/Frontend.php:
```diff
@@ -0,0 +1,428 @@
+	 */
+	protected function get_handler()  {
+
+		return $this->handler;
```
Fixed in https://github.com/skyverge/wc-plugin-framework/pull/507/commits/cdc548541c71d4ce242b2d76e00dbb29c519f5e2

deivamagalhaes commented on woocommerce/payment-gateway/External_Checkout/Google_Pay/Frontend.php:
```diff
@@ -0,0 +1,296 @@
+			'button_style'             => $this->get_handler()->get_button_style(),
+			'card_types'               => $this->get_handler()->get_supported_networks(),
+			'available_countries'	   => $this->get_handler()->get_available_countries(),
+			'currency_code'			   => get_woocommerce_currency(),
```
Fixed in https://github.com/skyverge/wc-plugin-framework/pull/507/commits/da144868ecb136de9c63ec6c14108ec736474383

deivamagalhaes commented:
> **Buy from Product Page**
> - [x] The shipping methods load as options, but the selected method doesn’t seem to come through as a line item & isn’t added to the total. The order ends up being the correct total w/ shipping

Fixed in https://github.com/skyverge/wc-plugin-framework/pull/507/commits/126bac7030d981517539b2218c063d5925cf36cc and https://github.com/skyverge/wc-plugin-framework/pull/507/commits/34b167f0c792b6823afcc2e85da7d513c1f30086.

> - [ ] Taxes are also not applied in the Google UI

I got taxes to be applied to the final price, but I am unsure how to add them to the initial price (before the user selects the shipping method) because we do not have an address at this point. We need an address to be able to calculate the taxes, right?

deivamagalhaes commented:
> **Virtual product from the Product page**
> - [x] The order is created with no line items and a $0 total

Fixed in https://github.com/skyverge/wc-plugin-framework/pull/507/commits/dae96e33cd9994bb106797ea44152795eaa5071c (the product wasn't being added to the cart).

- [x] A PHP notice is triggered: - There’s a notice: PHP Notice: Undefined property: stdClass::$shippingAddress in vendor/skyverge/wc-plugin-framework/woocommerce/payment-gateway/External_Checkout/Google_Pay/Google_Pay.php on line 438

Fixed in https://github.com/skyverge/wc-plugin-framework/pull/507/commits/4aba664e28ef6808503aabca7d4c850ce52be1d9 and https://github.com/skyverge/wc-plugin-framework/pull/507/commits/21771afe7ab3b9e168fd497b385b12f7832a9ceb.

There are still tax issues for virtual products as well.

deivamagalhaes commented:
> **Buy from Cart Page**
> - [x] If I change my shipping method in the Woo UI, the buttons disappear and don’t come back

Good catch! Fixed in https://github.com/skyverge/wc-plugin-framework/pull/507/commits/4a40970aa769e1e6f51b89929a46498dedf9c6d6.

> - [x] There appears to be an Apple Pay request logged (if enabled), but weirdly it logs as a Google request

Fixed in https://github.com/skyverge/wc-plugin-framework/pull/507/commits/8535aea3fc88fb8ec82453f5bbbb84199f846578.

deivamagalhaes commented:
@ChaseWiseman, please see the [above comment](https://github.com/skyverge/wc-plugin-framework/pull/507#issuecomment-717511890) about taxes.

deivamagalhaes commented:
> **Buy from Cart Page**
> - [x] If I change my shipping method in the Woo UI, the buttons disappear and don’t come back

Good catch! Fixed in https://github.com/skyverge/wc-plugin-framework/pull/507/commits/4a40970aa769e1e6f51b89929a46498dedf9c6d6 and https://github.com/skyverge/wc-plugin-framework/pull/507/commits/4568634449508e9f40c890b5809e689e8d7a7b86. I am wondering if there are other Ajax scenarios we need to consider.

> - [x] There appears to be an Apple Pay request logged (if enabled), but weirdly it logs as a Google request

Fixed in https://github.com/skyverge/wc-plugin-framework/pull/507/commits/8535aea3fc88fb8ec82453f5bbbb84199f846578.

deivamagalhaes commented:
@ChaseWiseman, it would be great if you could take another look at this one today.

ChaseWiseman commented:
@deivamagalhaes re-review already in progress 😄

ChaseWiseman reviewed (commented):
@deivamagalhaes most of the issues are looking fixed!

Still seeing this notice for virtual product orders:
```php
PHP Notice:  Undefined index: shippingAddress in ~/code/skyverge/woocommerce-gateway-cybersource/vendor/skyverge/wc-plugin-framework/woocommerce/payment-gateway/External_Checkout/Google_Pay/Google_Pay.php on line 471
```

>We need an address to be able to calculate the taxes, right?

Accurate taxes, yes. Does the API not allow for passing back basic billing info to get updated totals?

If not, then I'd recommend documenting what address is going to be used in each scenario so that limitation is known for the product and support teams. For instance, for each **Calculate tax based on** WC setting option, what address is going to be used for guest customers, logged in customers, shipping vs. virtual, etc...

Make sense?

deivamagalhaes commented:
> Accurate taxes, yes. Does the API not allow for passing back basic billing info to get updated totals?

If not using the dynamic price update flow, we only get the billing info after the payment is confirmed by the user. And I am not sure if we can use the dynamic price update flow if we are not using shipping. Let me test that and get back to you.

deivamagalhaes commented:
Yeah, I have confirmed we can't use the `onPaymentDataChanged` callback without shipping: https://developers.google.com/pay/api/web/reference/request-objects#PaymentDataCallbacks

When I try to hook it in, I get this error: 
```
If PaymentDataRequest.callbackIntents contains SHIPPING_ADDRESS, PaymentDataRequest.shippingAddressRequired must be set to true!
```

deivamagalhaes commented:
We get the billing information on the `onPaymentAuthorized` callback, but by then the customer already authorized the payment. What would happen is the user could see one value on the Google Pay pop-up and maybe be charged a different value if the tax calculation is different once we have the billing information.

This would be the behaviour for each scenario: 
- If "Calculate tax based on" is set to "Shop base address", I expect `wc_get_price_including_tax` to return the same price as the price with tax from the cart after the tax calculation
- If "Calculate tax based on" is set to "Customer shipping address" or "Customer billing address", and the product is shippable (non-virtual) I expect the price to be adjusted on the `onPaymentDataChanged` callback
- If "Calculate tax based on" is set to "Customer shipping address" or "Customer billing address", and the product is virtual, but the customer is logged in, I expect `wc_get_price_including_tax` to return the same price as the price with tax from the cart after the tax calculation
- If "Calculate tax based on" is set to "Customer shipping address" or "Customer billing address", and the product is virtual, but the customer is not logged in, and the tax rate for their billing address is different from the one for the shop address, we may have a small difference between the price displayed to the customer on the Google Pay pop-up (calculated by `wc_get_price_including_tax`) and the final priced charged to CyberSource

Where should we document that limitation?

deivamagalhaes commented:
Actually, I just tested the last scenario and that is not what I see. I see the tax rate for the shop address being applied regardless of the "Calculated tax based on" setting, for virtual products and guest customers. Maybe I need to recalculate taxes on the `onPaymentAuthorized` callback, after we set the billing address?

deivamagalhaes commented:
@ChaseWiseman, I am VERY confused about how this works...

Check this video:
https://cloud.skyver.ge/Z4uyGNA1

These are my settings:
- The product price (without taxes) is $37
- "Calculate tax based on" is set to "Customer billing address" 
![image](https://user-images.githubusercontent.com/4502503/97504934-30f11880-1935-11eb-81fc-040d9bda5e15.png)
- I have 2 tax rates:
![image](https://user-images.githubusercontent.com/4502503/97505028-4f571400-1935-11eb-960f-ac17491c7f54.png)
- My shop is based in CA
- The product is shippable and the customer is not logged in
- The Local pickup shipping option costs $5

This is what I got from the video:
1. The first price displayed on the Google Pay pop-up is $38.85, which is the price with CA tax rate applied (returned by `wc_get_price_including_tax` because we don't know the customer billing or shipping address)
1. The second price displayed on the Google Pay pop-up (after the `onPaymentDataChanged` callback) is $44.10, which corresponds to the product plus shipping, with the CA tax rate applied:
```
( $37 + $5 ) * ( 1.05 ) =  $44.10
```
So, in fact, the customer billing address is not being used at all on the tax calculation. I guess because it only looks for a CUSTOMER address, and it doesn't check the order address? Does that make sense? Is that because we don't have a `WP_User` for our customer (this would mean taxes never work properly for guest customers, it can't be the case). So what can we do to make sure `WC()->customer` is properly set? 🤔

deivamagalhaes commented:
We may need something similar to [this](https://github.com/skyverge/wc-plugin-framework/blob/master/woocommerce/payment-gateway/apple-pay/class-sv-wc-payment-gateway-apple-pay.php#L730)!

deivamagalhaes commented:
Damn it, I see the same prices even if I set the "Calculate tax based on" to "Customer shipping address". 😞

deivamagalhaes commented:
And still the same after implementing the same filter Apple Pay implements... This will require further debugging...

deivamagalhaes commented:
@ChaseWiseman , any idea why the `set_customer_taxable_address()` implementation only overwrites the address if the country is not the same as the base country? Is it assuming the tax rate is always the same for a whole country? 🤔

deivamagalhaes commented:
Ok, with the changes in https://github.com/skyverge/wc-plugin-framework/pull/507/commits/74b138704a9750ef38dc8871e05466a877da8ba1 I do see taxes calculated based on the shipping address, for shippable products.

deivamagalhaes commented:
Results seem consistent between Apple Pay and Google Pay for a shippable product: https://cloud.skyver.ge/mXu6GNQk

deivamagalhaes commented:
Also consistent for virtual products: https://cloud.skyver.ge/yAul9gRJ

No taxes are added either by Apple Pay or Google Pay.

Is this known limitation already documented for Apple Pay?

deivamagalhaes commented:
Also consistent for virtual products and taxes based on base (shop) address: https://cloud.skyver.ge/YEuy46n8

deivamagalhaes commented:
Summarizing the behaviour for each scenario:

- I have two tax rates, one for CA and one for PA
- my customer billing address is PA (on both Google Pay and Apple Pay)
- my shop address is CA

| Product type | Calculated tax based on | Customer | Shop state | Shipping state | Billing state | Tax rate applied |
| ------------- | ------------------------ | --------- |  ---------- | --------------- | ------------ | ---------------- |
| Shippable      | Shop base address         | Guest      | CA              | CA                     | PA                  | CA ✅                  | 
| Shippable      | Customer shipping address | Guest | CA              | CA                     | PA                  | CA ✅                  |   
| Shippable      | Customer shipping address | Guest | CA              | PA                     | PA                  | PA ✅                   |       
| Shippable      | Customer billing address | Guest     | CA              | CA                     | PA                  | CA ⚠️ (uses shipping state) |   
| Shippable      | Customer billing address | Guest     | CA              | PA                     | PA                  | PA ⚠️ (uses shipping state)  |   
| Virtual            | Shop base address         | Guest      | CA              | CA                     | PA                  | CA ✅                  | 
| Virtual            | Customer shipping address | Guest | CA              | N/A                    | PA                  | CA ⚠️ (uses shop state) | 
| Virtual            | Customer billing address | Guest    | CA              | N/A                    | PA                  | CA ⚠️ (uses shop state) | 

And the same behaviour happens for Apple Pay.

ChaseWiseman commented:
@deivamagalhaes thanks for summarizing those scenarios!

All of them seem to make sense based on the address info that's available to WC. For shippable products, it doesn't matter what the "Calculated tax based on" setting is, the only address available would be the shipping or base address.

For virtual products there is no address to use _except_ for the shop base address, right?

>We get the billing information on the onPaymentAuthorized callback, but by then the customer already authorized the payment. What I expect to happen is the user could see one value on the Google Pay pop-up and maybe be charged a different value if the tax calculation is different once we have the billing information.

Yes this is a concern - did you confirm whether this happens?

>So, in fact, the customer billing address is not being used at all on the tax calculation. I guess because it only looks for a CUSTOMER address, and it doesn't check the order address? Does that make sense? Is that because we don't have a WP_User for our customer (this would mean taxes never work properly for guest customers, it can't be the case). So what can we do to make sure WC()->customer is properly set?

It sounds like it's more due to the fact that we just don't have a billing address to use. If Google doesn't pass it back, and the guest or logged in user doesn't have a billing address, then there's nothing WC can use to calculate taxes but the base address. And there is no order at that point.

>Maybe we should display a notice on the Apple Pay/Google Pay settings page saying taxes can't be based on customer billing addresses?

We might want to, but I would definitely run your findings by the product team to ensure everyone is aware of the limitations, if in fact the assumptions above are true and there's nothing further we can do about them.

>Is this known limitation already documented for Apple Pay?

Not that I'm aware of - once the path is decided on how to handle this limitation in Google Pay, then we can open a story to apply that to Apple Pay for the backlog if needed.

deivamagalhaes commented:
@ChaseWiseman, please see my comments below. 🙂 

> @deivamagalhaes thanks for summarizing those scenarios! 

👍 

> All of them seem to make sense based on the address info that's available to WC. For shippable products, it doesn't matter what the "Calculated tax based on" setting is, the only address available would be the shipping or base address.

> For virtual products there is no address to use except for the shop base address, right?

Not before the payment is authorized.

>> We get the billing information on the onPaymentAuthorized callback, but by then the customer already authorized the payment. What I expect to happen is the user could see one value on the Google Pay pop-up and maybe be charged a different value if the tax calculation is different once we have the billing information.

> Yes this is a concern - did you confirm whether this happens?

It does not happen, whatever value they see on the pop-up is the value that is charged.

>> So, in fact, the customer billing address is not being used at all on the tax calculation. I guess because it only looks for a CUSTOMER address, and it doesn't check the order address? Does that make sense? Is that because we don't have a WP_User for our customer (this would mean taxes never work properly for guest customers, it can't be the case). So what can we do to make sure WC()->customer is properly set?

> It sounds like it's more due to the fact that we just don't have a billing address to use. If Google doesn't pass it back, and the guest or logged in user doesn't have a billing address, then there's nothing WC can use to calculate taxes but the base address. And there is no order at that point.

👍 

>> Maybe we should display a notice on the Apple Pay/Google Pay settings page saying taxes can't be based on customer billing addresses?

> We might want to, but I would definitely run your findings by the product team to ensure everyone is aware of the limitations, if in fact the assumptions above are true and there's nothing further we can do about them.

I do not believe there's anything we can do about the limitations. I'll send a message on the epic channel Slack with a link to the above table.

>> Is this known limitation already documented for Apple Pay?

> Not that I'm aware of - once the path is decided on how to handle this limitation in Google Pay, then we can open a story to apply that to Apple Pay for the backlog if needed.

Sounds good!

deivamagalhaes commented:
Slack conversation: https://skyverge.slack.com/archives/C01C5KDDV2P/p1604000769064600

deivamagalhaes commented:
@ChaseWiseman, @lgfogle will create a new story for the admin notices. So I believe we can merge this one to the release branch?

ChaseWiseman reviewed (approved):
@deivamagalhaes works for me!
